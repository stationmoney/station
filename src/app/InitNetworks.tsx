import { PropsWithChildren, useEffect, useState } from "react"
import axios from "axios"
import { ASSETS } from "config/constants"
import createContext from "utils/createContext"
import NetworkLoading from "./NetworkLoading"

export const [useNetworks, NetworksProvider] = createContext<{
  networks: InterchainNetworks
  filterEnabledNetworks: <T>(network: Record<string, T>) => Record<string, T>
}>("useNetworks")

const InitNetworks = ({ children }: PropsWithChildren<{}>) => {
  const [networks, setNetworks] = useState<InterchainNetworks>()
  const [enabledNetworks, setEnabledNetworks] = useState<string[]>([])

  useEffect(() => {
    const fetchChains = async () => {
      const { data: chains } = await axios.get<InterchainNetworks>(
        "/station/chains.json",
        {
          baseURL: ASSETS,
        }
      )
      setNetworks(chains)
    }

    fetchChains()
  }, [])

  useEffect(() => {
    const testChains = async () => {
      if (!networks) return
      const testBase = {
        ...networks.mainnet,
        ...networks.testnet,
        ...networks.classic,
      }

      const result = await Promise.all(
        Object.values(testBase).map(async (network) => {
          try {
            const { data } = await axios.get(
              "cosmos/base/tendermint/v1beta1/node_info",
              {
                baseURL: network.lcd,
                timeout: 5_000,
              }
            )
            return data.default_node_info.network as string
          } catch (e) {
            return null
          }
        })
      )

      setEnabledNetworks(
        result.filter((r) => typeof r === "string") as string[]
      )
    }

    testChains()
  }, [networks])

  if (!networks || !enabledNetworks.length)
    return <NetworkLoading title="Connecting to available networks..." />

  return (
    <NetworksProvider
      value={{
        networks,
        filterEnabledNetworks: (networks) =>
          Object.fromEntries(
            Object.entries(networks).filter(([chainID]) =>
              enabledNetworks.includes(chainID)
            )
          ),
      }}
    >
      {children}
    </NetworksProvider>
  )
}

export default InitNetworks
