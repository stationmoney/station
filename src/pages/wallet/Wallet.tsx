import styles from "./Wallet.module.scss"
import { ReactComponent as WalletCloseIcon } from "styles/images/icons/WalletCloseArrow.svg"
import CloseIcon from "@mui/icons-material/Close"
import { ReactComponent as BackIcon } from "styles/images/icons/BackButton.svg"
import { ReactComponent as WalletIcon } from "styles/images/menu/Wallet.svg"
import NetWorth from "./NetWorth"
import AssetList from "./AssetList"
import { useState } from "react"
import createContext from "utils/createContext"
import AssetPage from "./AssetPage"
import ReceivePage from "./ReceivePage"
import SendPage from "./SendPage"
import { atom, useRecoilState } from "recoil"
import TransferPage from "./TransferPage"

enum Path {
  wallet = "wallet",
  coin = "coin",
  receive = "receive",
  send = "send",
  transfer = "transfer",
}

type Route =
  | {
      path: Path.wallet
    }
  | {
      path: Path.coin
      denom: string
      previusPage: Route
    }
  | {
      path: Path.receive
      previusPage: Route
    }
  | {
      path: Path.send
      denom?: string
      previusPage: Route
    }
  | {
      path: Path.transfer
      denom?: string
      previusPage: Route
    }

// Handle routing inside Wallet
const [useWalletRoute, WalletRouter] = createContext<{
  route: Route
  setRoute: (route: Route) => void
}>("useWalletRoute")

export { useWalletRoute, Path }

export const isWalletBarOpen = atom({
  key: "isWalletBarOpen",
  default: true,
})

const Wallet = () => {
  const [isOpen, setIsOpen] = useRecoilState(isWalletBarOpen)
  const [route, setRoute] = useState<Route>({ path: Path.wallet })

  function BackButton() {
    if (route.path === Path.wallet) return null

    return (
      <button
        className={styles.back}
        onClick={() => setRoute(route.previusPage)}
      >
        <BackIcon width={18} height={18} />
      </button>
    )
  }

  function render() {
    switch (route.path) {
      case Path.wallet:
        return (
          <>
            <NetWorth />
            <AssetList />
          </>
        )
      case Path.coin:
        return (
          <>
            <BackButton />
            <AssetPage />
          </>
        )
      case Path.receive:
        return (
          <>
            <BackButton />
            <ReceivePage />
          </>
        )
      case Path.send:
        return (
          <>
            <BackButton />
            <SendPage />
          </>
        )
      case Path.transfer:
        return (
          <>
            <BackButton />
            <TransferPage />
          </>
        )
    }
  }

  return (
    <div className={`${styles.wallet} ${!isOpen && styles.wallet__closed}`}>
      <button
        className={styles.close}
        onClick={() => {
          setIsOpen((o) => !o)
        }}
      >
        {isOpen ? (
          <>
            <WalletCloseIcon
              width={18}
              height={18}
              className={styles.close__icon}
            />
            <CloseIcon
              width={18}
              height={18}
              className={styles.close__icon__mobile}
            />
          </>
        ) : (
          <>
            <span>Wallet</span>
            <WalletIcon width={16} height={16} />
          </>
        )}
      </button>
      <WalletRouter value={{ route, setRoute }}>{render()}</WalletRouter>
    </div>
  )
}

export default Wallet
