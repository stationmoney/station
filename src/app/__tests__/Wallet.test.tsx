import React from "react"
import { render } from "@testing-library/react"
import { RecoilRoot } from "recoil"
import { MemoryRouter } from "react-router-dom"
import Wallet from "pages/wallet/Wallet"
import { NetworksProvider } from "../InitNetworks"
import { mockExchangeRates } from "./__mocks__/ExchangeRates.mock"
import { mockNetworks } from "./__mocks__/Networks.mock"

function renderComponent() {
  type TokenFilter = <T>(network: Record<string, T>) => Record<string, T>
  const networks = {} as jest.Mocked<InterchainNetworks>
  const mockedTokenFilter = {} as jest.Mocked<TokenFilter>

  return render(
    <NetworksProvider
      value={{
        networks: networks,
        filterEnabledNetworks: mockedTokenFilter,
        filterDisabledNetworks: mockedTokenFilter,
      }}
    >
      <MemoryRouter>
        <RecoilRoot>
          <Wallet />
        </RecoilRoot>
      </MemoryRouter>
    </NetworksProvider>
  )
}

const mockCoins = [
  {
    denom:
      "ibc/BE2E170E02D101A1DC61255C51533FED2B6163FD2757DA11494FEB6D65ABEFC4",
    amount: "1000000000",
    chain: "pisco-1",
  },
  {
    denom: "uluna",
    amount: "1092273346",
    chain: "pisco-1",
  },
]

jest.mock("../../data/queries/bank", () => {
  const mockUseIsWalletEmpty = () => {
    return false
  }

  const mockUseBankBalance = () => {
    return mockCoins
  }

  return {
    useIsWalletEmpty: mockUseIsWalletEmpty,
    useBankBalance: mockUseBankBalance,
  }
})

jest.mock("../../data/wallet", () => {
  const mockUseNetworkName = () => {
    return "testnet"
  }

  const mockUseNetwork = () => {
    return mockNetworks
  }

  return {
    useNetworkName: mockUseNetworkName,
    useNetwork: mockUseNetwork,
  }
})

jest.mock("../../data/queries/coingecko", () => {
  const mockUseExchangeRates = () => {
    return mockExchangeRates
  }

  return {
    useExchangeRates: mockUseExchangeRates,
  }
})

jest.mock("../../data/token", () => {
  const actual = jest.requireActual("../../data/token")

  const mockUseNativeDenoms = () => {
    return () => {
      return { token: "uluna", decimals: 6, symbol: "LUNA" }
    }
  }

  return {
    ...actual,
    useNativeDenoms: mockUseNativeDenoms,
  }
})

describe("Wallet", () => {
  it("matches original component", () => {
    const { asFragment } = renderComponent()
    expect(asFragment()).toMatchSnapshot()
  })
})
