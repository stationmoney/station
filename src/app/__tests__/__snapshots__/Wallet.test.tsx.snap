// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet matches original component 1`] = `
<DocumentFragment>
  <div
    class="wallet false"
  >
    <button
      class="close"
    >
      <svg
        class="close__icon"
        height="18"
        width="18"
      >
        WalletCloseArrow.svg
      </svg>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium close__icon__mobile css-i4bv87-MuiSvgIcon-root"
        data-testid="CloseIcon"
        focusable="false"
        height="18"
        viewBox="0 0 24 24"
        width="18"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </button>
    <article
      class="networth"
    >
      <span
        class="inline start"
        style="gap: 4px;"
      >
        <p>
          Portfolio value
        </p>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit muted css-1vooibu-MuiSvgIcon-root"
          data-testid="HelpOutlineOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
          />
        </svg>
      </span>
      <h1>
        $ 
        <span
          class="component"
        >
          2,701
          <span
            class="small"
          >
            .27
          </span>
        </span>
      </h1>
      <div
        class="networth__buttons"
      >
        <button
          class="button primary"
          type="button"
        >
          Send
        </button>
        <button
          class="button default"
          type="button"
        >
          Receive
        </button>
      </div>
    </article>
    <article
      class="assetlist"
    >
      <div
        class="assetlist__title"
      >
        <h3>
          Assets
        </h3>
        <button
          class="item"
          type="button"
        >
          <span
            class="icon"
          />
          Manage tokens
        </button>
        <div />
      </div>
      <div
        class="assetlist__list"
      >
        <div>
          <section>
            <article
              class="asset"
            >
              <section
                class="details"
              >
                <img
                  alt=""
                  class="icon 50"
                  height="50"
                  src="https://assets.terra.money/icon/svg/Terra.svg"
                  style="padding: 14.285714285714286px;"
                  width="50"
                />
                <div
                  class="details__container"
                >
                  <h1
                    class="symbol"
                  >
                    LUNA
                    <span
                      class="chains"
                    >
                      Terra
                    </span>
                    <span
                      class="chains"
                    >
                      Terra
                    </span>
                    <span
                      class="chain__num"
                    >
                      2
                    </span>
                  </h1>
                  <h2
                    class="change__down"
                  >
                    <svg>
                      PriceDown.svg
                    </svg>
                     -2.81%
                  </h2>
                  <h1
                    class="price"
                  >
                    $ 
                    <span
                      class="component"
                    >
                      2,701
                      <span
                        class="small"
                      >
                        .27
                      </span>
                    </span>
                  </h1>
                  <h2
                    class="amount"
                  >
                    <span
                      class="component"
                    >
                      2,092
                      <span
                        class="small"
                      >
                        .273346
                      </span>
                    </span>
                     LUNA
                  </h2>
                </div>
              </section>
            </article>
          </section>
        </div>
      </div>
    </article>
  </div>
</DocumentFragment>
`;
